lib/Models/ArcGisMapServerCatalogItem.js

```diff
@@ -195,6 +203,43 @@ function getJson(item, uri) {
     return loadJson(proxyCatalogItemUrl(item, uri.addQuery('f', 'json').toString(), '1d'));
 }
 
+function getToken(item) {
+    const options = {
+        url: item.tokenSource,
+        method: 'POST',
+        headers: {'Content-Type': 'application/json'},
+        data: JSON.stringify({'url': item.url})
+    };
+
+    return loadWithXhr(options).then(function (result) {
+        const token = JSON.parse(result);
+
+        if (defined(token.token)) {
+            item._lastToken = token.token;
+            return token.token;
+        }
+
+        item._lastToken = undefined;
+        throw new RuntimeError('The token server responded with an invalid token.');
+    }).otherwise(() => {
+        item._lastToken = undefined;
+        throw new RuntimeError('Unable to request a token from the token server.');
+    });
+}
+
+function appendQuery(query, addition) {
```

Comment:
`urijs` has a handy method to do this: http://medialize.github.io/URI.js/docs.html#search-set

Reply:
Thanks for this.

f8b72f88960ec67655cc1b4805409459b3c0b78d